Initial license checkin

boomanaiden154 opened the pull request:
This patch adds license information for about 2000 packages from automated sources. The license information was obtained from Alpine Linux and PyPI and processed using tooling available in https://github.com/boomanaiden154/spack-license-utils. The license field was added in after all other directives in an automated fashion.

Note that while this license information is probably fairly accurate, it is not guaranteed to be accurate. In addition some of the license strings from Alpine Linux might not be valid SPDX license strings. Invalid SPDX identifiers can be picked up and fixed once we have validation/parsing infrastructure in place for the solver, and issues can be fixed as they come up.

boomanaiden154 commented:
@tgamblin 

I think one of the open questions with this change is if we want to split it up. I think doing it monolithically is easiest assuming everyone is fine with landing it soon (to avoid the significant possibility of many merge conflicts), but very open to changing things around if others have other suggestions.

alecbcs commented:
@boomanaiden154 it kinda looks like the license directive is all over in the rendered packages. If possible I think we should try to put it before any dependencies, versions, or variants. (Perhaps after the URL or maintainers directive as is shown in the new package template.)

boomanaiden154 commented:
> it kinda looks like the license directive is all over in the rendered packages. If possible I think we should try to put it before any dependencies, versions, or variants. (Perhaps after the URL or maintainers directive as is shown in the new package template.)

The scripts that I wrote just put it after all existing directives. So they should be after all of the other directives, but before everything else (or just at the end of the file if there isn't anything else). Should be easy enough to just put it before the version information like in the template.

boomanaiden154 commented:
@alecbcs I adjusted the formatting. Everything is now inserted above the `version` directive. I've also handled the edge cases I've run across like the `version` field being sandwiched between other variable declarations/directives and when it is inside control flow. Hopefully things look better now.

alecbcs commented on var/spack/repos/builtin/packages/3proxy/package.py:
```diff
@@ -14,6 +14,8 @@ class _3proxy(MakefilePackage):
     homepage = "https://3proxy.org"
     url = "https://github.com/z3APA3A/3proxy/archive/0.8.13.tar.gz"
 
+    license("Apache-2.0 GPL-2.0-or-later LGPL-2.1-or-later")
```
```suggestion
    license("Apache-2.0", "GPL-2.0-or-later", "LGPL-2.1-or-later")
```

I may have missed this in the initial PR, but I thought multiple licenses would be declared like multiple maintainers? (Each license would be its own string in a series of parameters to the license directive.)

boomanaiden154 replied:
They're supposed to be valid [SPDX expressions](https://spdx.github.io/spdx-spec/v2.3/SPDX-license-expressions/). The `license` directive only takes in a single license expression and an optional when spec, and they all have to match mutually exclusive specs. There are a couple expressions that aren't valid at the moment (including this one) as it should be using the `OR` operator. To properly validate these, we need a compliant expression parser. I'm planning on implementing one in the near future, but not sure when I'll be able to get to it.

The plan (on my end) was as follows:
1. Get initial automatically collected license information in (this PR).
2. Add in a SPDX expression parser.
3. Enable the use of the expression parser for validating package license expressions/fix existing invalid expressions.
4. More work to better integrate licenses into Spack.

Landing this in some form (or I guess I could just pull the commit into my fork...) unblocks me on a couple downstream tasks and would also enable large scale testing of a couple other patches I have planned, such as adding license information to https://packages.spack.io.

boomanaiden154 replied:
Most of these issues should be fixed now with my latest update, more details below.

boomanaiden154 commented:
A couple changes in the latest revision
* Pulled license information in from more sources. I added in information from CRAN and used the `go-license-detector` on package sources downloaded with `spack stage`. That puts as at about ~5400/7500 packages tagged with license information. I would hypothesize that more exploration with automated license annotating is now going to see diminishing returns.
* Added in basic linting support. I essentially just strip most of the syntax from SPDX expressions and then make sure the license IDs aren't valid. This isn't perfect,  but should catch most of the egregious issues.

boomanaiden154 commented:
In the latest update I've added in a comment on all of the automatically generated license annotations noting that they're automatically generated so one doesn't have to go to the git blame to figure that out and to encourage manual verification. Happy to change up the approach there if there are qualms with that though.

haampie commented:
Maybe also reduce diff size a bit with a one line comment instead of 3? ;p

boomanaiden154 commented:
> Maybe also reduce diff size a bit with a one line comment instead of 3? ;p

Do you have a concise phrasing that you would prefer? I wanted to make the message fairly readable with all the necessary information and flake8 complains about line widths if they're longer than ~80 characters.

haampie commented:
```
# NOTE: license needs to be verified, see: https://github.com/spack/spack/issues/12345
```

or something?

boomanaiden154 commented:
> ```
> # NOTE: license needs to be verified, see: https://github.com/spack/spack/issues/12345
> ```
> 
> or something?

Done. Modified the message slightly to make it a bit shorter to fit.

tgamblin commented:
@boomanaiden154 wiuldnit be possible to have, say, a `spack license verify <name>` package that downloads tarballs and runs `licensee`?  With that automation we’d have enough to check new versions of packages as they are added in PRs.

boomanaiden154 commented:
> @boomanaiden154 would it be possible to have, say, a `spack license verify <name>` package that downloads tarballs and runs `licensee`? With that automation we’d have enough to check new versions of packages as they are added in PRs.
> 
> Is it worth looking at how well that works across these 2k packages?

It would be possible, but I think there are probably enough edge cases in license detection where it would be fairly noisy without additional annotations specifying where to look for license information/a significant amount of engineering effort. In some of the more complicated cases that I've seen there might be one top level license file and then several subfolders, only a couple of which are actually built from, with each subfolder using a different license. Those are edge cases and aren't that common, but they might just be frequent enough to be annoying.

The tool that I'm using currently to do the license detection, `go-license-detector`, would also probably require some packaging work to enable use from within spack as it is a go package.

A naive technique that just scans I assume would validate most of these licenses, but not catch the edge cases where they aren't. I'm already using license detection to generate something like ~2.5k out of the ~5k license annotations added here, so using the same technique would of course "validate" those.

Also, is there a timeline for when we'd be able to get this merged? The merge commits aren't too bad to deal with (especially when just retagging all the license information), but they do stack up pretty quickly.

boomanaiden154 committed:
Initial License Checkin

This patch adds license information for about 5,300 packages from automated sources.
The license information was obtained from Alpine Linux and PyPI and processed
using tooling available in https://github.com/boomanaiden154/spack-license-utils.
The license field was added in after all other directives in an automated fashion.

Note that while this license information is probably fairly accurate, it is not
guaranteed to be accurate. In addition some of the license strings from Alpine Linux
might not be valid SPDX license strings. Invalid SPDX identifiers can be picked up
and fixed once we have validation/parsing infrastructure in place for the solver,
and issues can be fixed as they come up.

tgamblin committed:
directives: add `checked_by` field to `license()`, add some license checks

Add a "checked_by" field to the `license()` directive so that we can track who verified
the license for a project. also check the license of 18 or so projects and mark them
checked.

tgamblin commented:
@boomanaiden154: I removed the "this needs verification lines" and instead added a `checked_by=<username>` argument to `license()` so that we can mark the checked ones without quite as much noise.  I checked 18 or so of the more complicated ones with AND.

I think I may wan to factor out AND/OR/WITH so that we can more easily list licenses on multiple lines, based on brew's license guidelines: 
* https://docs.brew.sh/License-Guidelines

Also, we might want to add some location where we can record new valid license ids and URLs to what they are, based on:
* https://github.com/NixOS/nixpkgs/blob/master/lib/licenses.nix

boomanaiden154 commented:
> @boomanaiden154: I removed the "this needs verification lines" and instead added a `checked_by=<username>` argument to `license()` so that we can mark the checked ones without quite as much noise. I checked 18 or so of the more complicated ones with AND.

Sounds good. I think in terms of raw diff counts, it would be about the same. Having something explicit in the directive that the license has been validated will be nice though. 

> For now I enabled auto-merge.

Looks like the Gitlab CI pipeline originally failed and then merging in `develop` cancelled that, so it will be enabled again. Sorry for the churn there.

> For later PRs:
> 
> I think I may wan to factor out AND/OR/WITH so that we can more easily list licenses on multiple lines, based on brew's license guidelines:
> 
> * https://docs.brew.sh/License-Guidelines

We can do that. I wouldn't be opposed to sticking with just SPDX expression strings as they are specified, already used widely, and should be just as expressive without having to reinvent the wheel.

> Also, we might want to add some location where we can record new valid license ids and URLs to what they are, based on:
> 
> * https://github.com/NixOS/nixpkgs/blob/master/lib/licenses.nix

Yep. Having a spot in the documentation to do this would be a good idea. I'd like to coordinate with upstream SPDX as much as possible though to avoid divergence there. They do add common licenses each release and do a new release quarterly. If there are any major gaps we observe, I should be able to hop on the SPDX meetings and figure out what we can do there. There will be a couple things like the `CUSTOM` license for things that don't conform to any common license that should be in there too.

alecbcs commented:
@spackbot run pipeline

tgamblin commented:
Ok trying this again with a few more licenses checked.  I am not wed to the `checked_by="github_login"` thing but it seemed concise and like a reasonable way to indicate who you're trusting when you believe your license attestations.  It's not really a formal attestation though (or a guarantee the maintainer knows anything about licenses -- though hopefully we do). It at least:
1. Requires someone to positively indicate that the license is valid (vs. casually ignoring it), which I think is worth something
2. Keeps things to one line.

tgamblin commented:
There is one failure here (PAPI on Cray), but I confirmed it's broken on `develop`.  I'll work on that separately and merge this.

tgamblin commented:
Thanks for all the work on licenses @boomanaiden154!
